### PR TITLE
Add RAJA::synchronize

### PR DESCRIPTION
**Pull request by davidbeckingsale**

### CODE REVIEW

File: include/RAJA/policy/openmp/synchronize.hpp [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+{
+
+/*!
+ * \brief Synchronize all OpenMP threads and tasks.
+ */
+RAJA_INLINE
+void
+synchronize_impl(const omp_synchronize&)
+{
+#pragma omp barrier
```
Comment (trws) [inline include/RAJA/policy/openmp/synchronize.hpp]: I just double checked this, the barrier is sufficient to wait for all tasks, so the taskwait can come out unless we want a slightly smaller hammer.

File: test/unit/cpu/test-synchronize.cpp [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+//
+// For details about use and distribution, please read RAJA/LICENSE.
+//
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "RAJA/RAJA.hpp"
+#include "gtest/gtest.h"
+
+TEST(SynchronizeTest, omp){
+  RAJA::synchronize<RAJA::omp_synchronize>();
```
Comment (trws) [inline test/unit/cpu/test-synchronize.cpp]: Maybe add something that requires a sync here? A nanosleep(10) with a set or something?

File: include/RAJA/policy/openmp/synchronize.hpp (hunk 1 above)
Comment (davidbeckingsale) [reply include/RAJA/policy/openmp/synchronize.hpp]: Great, the `taskwait` fails on AppVeyor too.

File: test/unit/cpu/test-synchronize.cpp [hunk 3]
```diff
@@ -16,6 +16,22 @@
...
 #include "gtest/gtest.h"
 
+#if defined(RAJA_ENABLE_OPENMP)
+
 TEST(SynchronizeTest, omp){
-  RAJA::synchronize<RAJA::omp_synchronize>();
+
+  double test_val = 0.0;
+
+#pragma omp parallel shared(test_val)
```
Comment (davidbeckingsale) [inline test/unit/cpu/test-synchronize.cpp]: @trws - this okay?

Comment (trws) [reply test/unit/cpu/test-synchronize.cpp]: Sure, that'll work.

Comment (davidbeckingsale) [reply test/unit/cpu/test-synchronize.cpp]: 👍

File: include/RAJA/policy/cuda.hpp [hunk 4]
```diff
@@ -42,6 +42,7 @@
 #if defined(__NVCC__)
```
Comment (rhornung67) [inline include/RAJA/policy/cuda.hpp]: Why is this guard needed here? The contents of the file are guarded by RAJA_ENABLE_CUDA and the other policy/cuda/*.hpp inclusions are not guarded. If the guard is actually needed, I think it should go inside the header.

Comment (davidbeckingsale) [reply include/RAJA/policy/cuda.hpp]: That was added by @willkill07 I think - nothing to do with this PR.

File: test/unit/cuda/test-synchronize.cpp [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+// For details about use and distribution, please read RAJA/LICENSE.
+//
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+
+#include "RAJA/RAJA.hpp"
+#include "gtest/gtest.h"
+
+TEST(SynchronizeTest, CUDA){
+  RAJA::synchronize<RAJA::cuda_synchronize>();
```
[the remark this answers is not in the document]
Comment (davidbeckingsale) [reply test/unit/cuda/test-synchronize.cpp]: I was mainly testing it compiles 😄

**Comment by davidbeckingsale** [comment]
@ajkunen - I added a real test

File: test/unit/cuda/test-synchronize.cpp [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  double* managed_data;
+  cudaMallocManaged(&managed_data, sizeof(double)*50);
+
+  RAJA::forall<RAJA::cuda_exec_async<256>>(0,50, [=] RAJA_DEVICE (RAJA::Index_type i) {
+      managed_data[i] = 1.0*i;
+  });
+  RAJA::synchronize<RAJA::cuda_synchronize>();
+
+  RAJA::forall<RAJA::loop_exec>(0,50, [=] (RAJA::Index_type i) {
+      EXPECT_EQ(managed_data[i], 1.0*i);
```
Comment (ajkunen) [inline test/unit/cuda/test-synchronize.cpp]: wonderful

File: include/RAJA/policy/cuda/policy.hpp [hunk 7]
```diff
@@ -196,17 +196,27 @@ static_assert(MAX_BLOCK_SIZE % WARP_SIZE == 0,
...
+    : make_policy_pattern_launch_t<Policy::cuda, Pattern::synchronize, Launch::sync> {
+};
+
 }  // end namespace cuda
 }  // end namespace policy
 
 using policy::cuda::cuda_exec;
+
+template<size_t BLOCK_SIZE>
+using cuda_exec_async = policy::cuda::cuda_exec<BLOCK_SIZE, true>;
```
Comment (davidbeckingsale) [inline include/RAJA/policy/cuda/policy.hpp]: @trws @ajkunen I added an alias for `cuda_exec_async` also. The boolean argument is a bit opaque.

**Pull request merged**